echo-rest/src/main/java/com/netflix/spinnaker/echo/events/RestEventListener.java

```diff
@@ -55,79 +68,188 @@ class RestEventListener implements EventListener {
-                  log.error(
-                      "Could not send event source={}, type={} to {}.\n Event details: {}",
-                      event.getDetails().getSource(),
-                      event.getDetails().getType(),
-                      service.getConfig().getUrl(),
-                      event,
-                      e);
-                } else {
-                  log.error("Could not send event.", e);
-                }
-
-                registry
-                    .counter("event.send.errors", "exception", e.getClass().getName())
-                    .increment();
-              }
-            });
+    for (RestUrls.Service service : restUrls.getServices()) {
+      if (circuitBreakerEnabled) {
+        processEventWithCircuitBreaker(event, service);
+      } else {
+        processRestEvent(event, service);
+      }
+    }
+  }
+
+  /**
+   * Processes the incoming event and sends it to the specified service using REST.
+   *
+   * @param event The event to be processed and sent.
+   * @param service The service to which the event is sent.
+   */
+  private void processRestEvent(Event event, RestUrls.Service service) {
+    Map<String, Object> eventMap = transformEventToMap(event, service);
+    try {
+      sendEvent(eventMap, service);
+    } catch (Exception e) {
+      handleError(event, service, e);
+    }
+  }
+
+  /**
+   * Processes the incoming event with a {@link CircuitBreaker}, decorating {@link
+   * #processEvent(Event)}
+   *
+   * <p>Your service event name {@link RestProperties.RestEndpointConfiguration#getEventName()}
+   * should map to the Circuit Breaker instance name {@link CircuitBreaker#getName()}
+   *
+   * <p>Example config:
+   *
+   * <pre>{@code
+   * resilience4j.circuitbreaker:
+   *   instances:
+   *     sendEvent:
+   *       slow-call-duration-threshold: 10s
+   *       failure-rate-threshold: 40
+   *       minimum-number-of-calls: 2
+   *       wait-duration-in-open-state: 30s
+   *       permitted-number-of-calls-in-half-open-state: 1
+   * rest:
+   *   enabled: true
+   *   circuit-breaker-enabled: true
+   *   endpoints:
+   *     - url: "https://my-service.com"
+   *       eventName: "sendEvent"
+   * }</pre>
+   *
+   * @param event The event to be processed and sent.
+   * @param service The service to which the event is sent.
+   */
+  private void processEventWithCircuitBreaker(Event event, RestUrls.Service service) {
+    String circuitBreakerInstance =
+        StringUtils.defaultString(service.getConfig().getEventName(), "default");
+
+    CircuitBreaker circuitBreaker = circuitBreakerRegistry.circuitBreaker(circuitBreakerInstance);
+
+    try {
+      // acquire permission before writing the event to a Map
+      // this will save processing and memory usage if the Circuit Breaker is OPEN or HALF_OPEN
+      circuitBreaker.acquirePermission();
+
+      Map<String, Object> eventMap = transformEventToMap(event, service);
```

Comment:
if an exception is thrown in the `transformEventToMap` method. does it try to send an event with errors?

Reply:
https://github.com/spinnaker/echo/pull/1315#discussion_r1283411288